workflow/rules/mapping.smk

```diff
@@ -0,0 +1,66 @@
+rule bwa_index:
+    input:
+        config["resources"]["ref"]["genome"]
+    output:
+        multiext(config["resources"]["ref"]["genome"], ".amb", ".ann", ".bwt", ".pac", ".sa")
+    log:
+        "logs/bwa/bwa_index.log"
+    params:
+        algorithm="bwtsw"
+    wrapper:
+        "0.55.1/bio/bwa/index"
+
+rule bwa_mem:
+    input:
+        reads = get_map_reads_input,
+        idx = rules.bwa_index.output
+    output:
+        temp("results/mapped/{sample}-{unit}.bam")
+    log:
+        "logs/bwa/bwa_mem/{sample}-{unit}.log"
+    params:
+        index= lambda w, input: os.path.splitext(input.idx[0])[0],
+        extra= get_read_group,
+        sort="samtools",
+        sort_order="coordinate",
+    threads: 8
+    wrapper:
+        "0.52.0/bio/bwa/mem"
+
+rule mark_duplicates:
+    input:
+        rules.bwa_mem.output
+    output:
+        bam=temp("results/mapped/dedup/{sample}-{unit}.bam"),
+        metrics="results/mapped/dedup/{sample}-{unit}.metrics.txt"
+    log:
+        "logs/picard/dedup_{sample}-{unit}.log"
+    params:
+        "REMOVE_DUPLICATES=true"
+    wrapper:
+        "0.55.1/bio/picard/markduplicates"
+
+rule merge_bams:
+    input:
+        get_dedup_bams
+    output:
+        temp("results/mapped/merged_with_dups.bam")
```

Comment:
I think we don't actually need a separately defined input function, here. And what we want to do is merge all the `unit`s of a `sample` into one file, so reducing / aggregating from `{sample}-{unit}` combinations down to one file per `{sample}`:
```
    input:
        lambda w: expand("results/mapped/dedup/{sample}-{unit}.bam",
            sample = w.sample,
            unit = units.loc[units['sample'] == w.sample].unit.to_list()
        )
    output:
        temp("results/merged/{sample}.bam")
```
I think, the previous version of the code would merge all samples into one BAM file, right? This now goes and looks up all the units that belong to each sample and requests the respective BAM files as input to merge them into one sample-specific BAM file. Also, this wades into [`pandas` territory](https://pandas.pydata.org/pandas-docs/stable/user_guide/index.html) with a syntax that often has many alternatives which I often find confusing. But is also quite useful, so it's worth sifting through the docs and Stackoverflow (see e.g. [the many ways of selecting a row in a pandas data frame](https://stackoverflow.com/a/46165056)...).